### DISCUSSION TITLE
既存のマネキン準拠のキャラクターで認証を通したいです

### DISCUSSION THREAD

**Discussion by papicochan** [pixiv/vroid-sdk-developers#177] [Q&A] [answered]
https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e

Noteの 既存のMannequin準拠のCharacterをVRMに差し替える の項を読んで、デフォルトのThirdpersoncharacterにD&Dでモデルを差し替える機構を追加し、使用させて頂いておりました。

ここに認証を通すパターンも追加する事は難しいでしょうか。

元々SDKの中に用意して頂いているキャラクターを使えばいいところなのですが、UE自体初心者の私の設計が悪く、Thirdpersoncharacterの方に変数や処理の書き込みを多くしてしまっており、コピーする事が難しく、なんとか今使っているThirdpersoncharacterBP(サードパーソンテンプレートに最初から入っているキャラクターのBP)に認証を通したパターンを移植できると有り難いです。

認証レベルを開いた時に、Hubを使用するというブーリアンをオンにし、Thirdpersoncharacterのロード時に該当のブーリアンがオンであれば認証を通したパターンんで差し替え...という道を考えておりましたがシステムに関する理解が浅くうまく行きませんでした。

**Comment by mameoPix** [comment]
お問い合わせいただきありがとうございます。
問題解決に向けて、出来る限りサポートさせていただきますので、どうぞよろしくお願いいたします。

つきましては、いくつか開発環境についてお伺いできればと思います。
お手数ですが、使用されているUnreal EngineのバージョンやSDKのバージョン、可能であればVRM4Uのバージョンもご教示いただけますでしょうか？

以下に例を記載いたします：
Unreal Engine：5.3.2
VRoid SDK：1.2.0
VRM4U：VRM4U_5_3_20240823

**Reply by papicochan** [reply]
お世話になります。
暖かいお言葉とサポートいつも有難うございます。

Unreal Engine 5.3.2
VroidSDK1.1.0
VRM4U 5_3_20231126

上記の組み合わせで使用しております。
最新版の利用ではなく申し訳ないのですがよろしくお願い致します。
(最新版に差し替えても動作に問題がないか検証中です)

**Reply by mameoPix** [reply]
開発環境について共有いただき、ありがとうございます！
ご提示いただいた SDK 1.1.0 と VRM4U 5_3_20231126 の組み合わせですが、下記の図のようなエラーが発生し、プロジェクトが正常に起動できないかと思います。
![image](https://github.com/user-attachments/assets/b1fed9a8-c2a5-411f-8c53-5f9298f79612)
というのも、SDK1.1.0で使用しているVRM4Uの関数が異なってくるためです。
https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e?pvs=25#3d92884c2d374593bad2ffb606f19af8
そのため、上記ドキュメントの**Pluginの導入方法**の箇所に以下のような記載をしております。
> VRoid SDK for Unreal Engine 0.5.0 以上を利用される場合、VRM4Uは20240402以上を利用してください。

大変お手数をおかけしてしまって申し訳ございませんが
問題を解決するために、まずSDKとVRM4Uのバージョンを更新してみていただけないでしょうか？
こちらでも再現環境を揃えて検証させていただきます。

更新手順は以下の通りです
1. プロジェクトが起動しなくなる / 実行時にエラーが発生して意図した挙動にならなくなる可能性もあるため、まずプロジェクトのバックアップ（プロジェクトをコピーするなど）を取ってください。
2. プロジェクトの Plugins フォルダ内にある VRoid SDK と VRM4U のフォルダを削除します。
3. 新しく取得したSDKおよびVRM4Uを Plugins フォルダに入れます。
4. プロジェクトを再度起動します。

お手数をおかけしますが、どうぞよろしくお願いいたします。

**Reply by mameoPix** [reply, marked as answer]
肝心の質問に対する回答ですが
認証レベルからThirdPersonMap等のゲームレベルへの流れにおいて
認証レベルで選択したVRMをThirdPersonMapで使用する手順については
独自のCharacter Blueprintに以下の設定を追加していただくだけで、変更が可能となっております。
![image](https://github.com/user-attachments/assets/2e9eb9c5-5865-45fb-bf78-a56f6a42da2b)

認証Levelで使用しているPreview用のCharacter自体を差し替えるとなると
`BP_AuthActor`内のEventGraphにある`Event OnDownloadVrm`内で使用している
BP_VRoidMannequinRetargetCharacterを独自のCharacter Blueprintに差し替えた後に
Update VRMノードと同様の処理を組む必要があります。
![image](https://github.com/user-attachments/assets/a63421c7-9107-4d6a-9637-931730713642)

**Reply by papicochan** [reply]
ご丁寧にありがとうございます
実現したかったものが1枚目の方で無事実装できました！

プラグインの組み合わせですが、何故か私の環境ではエラーなくビルドできてしまっていて気付いておりませんでした。

念の為プラグインの更新もできまして有難かったです。
本当にありがとうございました！
